### PR TITLE
Asset

### PR DESCRIPTION
**Pull request by fergusdixon**
## Overview

| Q                 | A
| ----------------- | ---
| Bugfix?           | no
| New feature?      | yes
| Breaking Changes? | no
| Fixed issues      | #

## The problem
No Asset

## How I resolved it
Made asset

## How to verify it
Have a look

Notify the following people: @darrynten

### CODE REVIEW

File: tests/SageOne/Models/AssetCategoryModelTest.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        $this->verifyGetUndefined(AssetCategory::class);
+    }
+
+    public function testCanNotNullify()
+    {
+        $this->verifyCanNotNullify(AssetCategory::class, 'id');
+    }
+
+   
+
```
Comment (darrynten) [inline tests/SageOne/Models/AssetCategoryModelTest.php]: Too many newlines here

File: tests/SageOne/Models/AssetCategoryModelTest.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        $this->verifyFeatures(AssetCategory::class, [
+            'all' => true, 'get' => true, 'delete' => true, 'save' => true
+        ]);
+    }
+
+    public function testGetAll()
+    {
+        $this->verifyGetAll(AssetCategory::class, function ($results) {
+            $this->assertEquals(2, count($results));
+            $this->assertEquals("sample string 1", $results[0]->description);
```
Comment (darrynten) [inline tests/SageOne/Models/AssetCategoryModelTest.php]: We use single quotes for these strings

File: tests/SageOne/Models/AssetCategoryModelTest.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            $this->assertEquals(2, $results[0]->id);
```
Comment (darrynten) [inline tests/SageOne/Models/AssetCategoryModelTest.php]: If there's more than 1 in the array we should check both

File: tests/SageOne/Models/AssetCategoryModelTest.php [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            $this->assertEquals('2017-07-17', $results[0]->modified->format('Y-m-d'));
+            $this->assertEquals('2017-07-17', $results[0]->created->format('Y-m-d'));
+        });
+    }
+    public function testGetId()
```
Comment (darrynten) [inline tests/SageOne/Models/AssetCategoryModelTest.php]: We need to leave a newline between the end of the last method and the start of this method

File: tests/SageOne/Models/AssetCategoryModelTest.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            $this->assertEquals("sample string 1", $model->description);
+            $this->assertEquals(2, $model->id);
+            $this->assertEquals('2017-07-17', $model->modified->format('Y-m-d'));
+            $this->assertEquals('2017-07-17', $model->created->format('Y-m-d'));
+        });
+    }
+    public function testSave()
+    {
+        $this->verifySave(AssetCategory::class, function ($model) {
+            $model->description = "sample string 1";
```
Comment (darrynten) [inline tests/SageOne/Models/AssetCategoryModelTest.php]: Single quotes here too :)

File: tests/SageOne/Models/AssetLocationModelTest.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        $this->verifyGetUndefined(AssetLocation::class);
+    }
+
+    public function testCanNotNullify()
+    {
+        $this->verifyCanNotNullify(AssetLocation::class, 'id');
+    }
+
+   
+
```
Comment (darrynten) [inline tests/SageOne/Models/AssetLocationModelTest.php]: Too many newlines here

File: tests/SageOne/Models/AssetLocationModelTest.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    public function testGetId()
+    {
+        $this->verifyGetId(AssetLocation::class, 2, function ($model) {
+            $this->assertInstanceOf(AssetLocation::class, $model);
+            $this->assertEquals("sample string 2", $model->description);
+            $this->assertEquals(1, $model->id);
+        });
+    }
+    public function testSave()
```
Comment (darrynten) [inline tests/SageOne/Models/AssetLocationModelTest.php]: More newlines between end and beginning of methods

File: tests/SageOne/Models/AssetModelTest.php [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+            $this->assertEquals('2017-07-17', $results[0]->category->created->format('Y-m-d'));
+            $this->assertInstanceOf(AssetLocation::class, $results[0]->location);
+            $this->assertEquals(1, $results[0]->location->id);
+            $this->assertEquals("sample string 2", $results[0]->location->description);
+            $this->assertEquals('2017-07-17', $results[0]->datePurchased->format('Y-m-d'));
+            $this->assertEquals("sample string 3", $results[0]->serialNumber);
+            $this->assertEquals("sample string 4", $results[0]->boughtFrom);
+            $this->assertEquals(1.0, $results[0]->purchasePrice);
+            $this->assertEquals(1.0, $results[0]->currentValue);
+            $this->assertEquals(1.0, $results[0]->replacementValue);
```
Comment (darrynten) [inline tests/SageOne/Models/AssetModelTest.php]: We should also do checks against the second result

File: tests/SageOne/Models/AssetModelTest.php [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+            $this->assertInstanceOf(Asset::class, $model);
+            $this->assertEquals("sample string 1", $model->description);
+            $this->assertInstanceOf(AssetCategory::class, $model->category);
+            $this->assertEquals("sample string 1", $model->category->description);
+            $this->assertEquals(2, $model->category->id);
+            $this->assertEquals('2017-07-17', $model->category->modified->format('Y-m-d'));
+            $this->assertEquals('2017-07-17', $model->category->created->format('Y-m-d'));
+            $this->assertInstanceOf(AssetLocation::class, $model->location);
+            $this->assertEquals(1, $model->location->id);
+            $this->assertEquals("sample string 2", $model->location->description);
```
Comment (darrynten) [inline tests/SageOne/Models/AssetModelTest.php]: Single quotes here

**Review by darrynten — CHANGES_REQUESTED** [review]
Some minor change requests, looking good though!

File: tests/SageOne/Models/AssetCategoryModelTest.php (hunk 3 above)
Comment (fergusdixon) [reply tests/SageOne/Models/AssetCategoryModelTest.php]: There was only 1 ID, set to 2
`{
  "TotalResults": 1,
  "ReturnedResults": 2,
  "Results": [
    {
      "Description": "sample string 1",
      "ID": 2,
      "Modified": "2017-07-17",
      "Created": "2017-07-17"
    },
    {
      "Description": "sample string 1",
      "ID": 2,
      "Modified": "2017-07-17",
      "Created": "2017-07-17"
    }
  ]
}`

Comment (darrynten) [reply tests/SageOne/Models/AssetCategoryModelTest.php]: Oh my... It looks like all of Sages' mocks return the same data for each item in a list 🤔

Comment (fergusdixon) [reply tests/SageOne/Models/AssetCategoryModelTest.php]: Yea I'm not 100% sure of the duplication in `$results` for the getAll methods in the mocks

**Review by darrynten — APPROVED** [review]

**Pull request merged**